design/host-discovery.md

```diff
@@ -0,0 +1,292 @@
+state. The state machine for the `baremetal-operator` will therefore
+need a new transition between `Discovered` and `MatchProfile` when the
+condition `!externallyProvisioned && !NeedsHardwareInspection()` is
+true. That transition will bypass the `Inspecting` state and allow the
+host to eventually move to the `Ready` state.
+
+When a user adds BMC credentials to a host in the `Discovered` state
+today it moves to `Registering` and the `Provisioner` is responsible
+for configuring the host in Ironic. That logic may need to be modified
+to support partial updates of existing nodes, to set the values that
+could not be discovered.
+
+There is a race between a server booting and a host resource for it
+being created. Today if the server boots before the host resource
+exists and the node is registered in Ironic, the host resource is
+effectively broken and the user must clean up both the resource and
+Ironic manually. Supporting discovery means we can fix the race by
+allowing a user-created host resource to be matched to an existing
+node in Ironic. The only real piece of information we have for making
+the match is the MAC address, but that should be a sufficiently unique
+identifier. The `IronicProvisioner` needs to be updated to recognize
+when a node with the same MAC already exists. If the node has no name,
+indicating that it was discovered, the provisioner should set the name
+and update any other values not set through discovery. If the node
+does have a name and the name does not match the name of the
+`BareMetalHost` then the host resource being reconciled should enter
+the `Error` state and an error message should be attached explaining
+that it is a duplicate host.
+
+### Implementation Details/Notes/Constraints
+
+The state management changes described above are likely to require
+changes to the `Provisioner` API. The details of those changes will be
+determined during implementation.
+
+The changes to the state machine would not change the fact that many
+of the metal3 features (provisioning and power management) require BMC
+credentials. Discovered hosts would not enter the states where those
+features are allowed until BMC credentials are provided. The host
+controller should skip trying to maintain the power state and should
+not honor reboot requests for hosts without BMC credentials.
+
+### Risks and Mitigations
+
+Data on discovered hosts will be lost if automatic provisioning is
+enabled. The feature is disabled by default, so the risk is mitigated
+because users will be aware when it is turned on.
+
+Ironic inspector creates discovered nodes with an `auto_discovered`
+flag set in the introspection data. We could limit the polling in the
+`baremetal-operator` to only look for hosts with the flag set. It
+isn’t really clear how useful that would be, so the description above
+does not assume such a limit.
+
+There does not appear to be a way to remove inspection data from
```

Comment:
Yes, here I had been considering the BMO looking at the ironic-inspector API rather than the ironic API, and I was trying to work out how we could tell the difference between a host discovered by ironic but deleted from the cluster and one that we actually want to bring into the cluster.

If we end up with a host in k8s that the user doesn't want, its data will be deleted from ironic when the host CR is removed. If the same host posts discovery info again, the same records will be re-created. I don't think there's a way to avoid that, other than not booting the discovery image on the host.

Maybe this bit is a digression that we don't need in this design doc.